doc/sphinx/proof-engine/vernacular-commands.rst

```diff
@@ -220,63 +220,79 @@ Requests to the environment
+      | {? - } @one_term
+
+   This command displays the name and type of objects (hypothesis of
+   the current goal, theorems, axioms, etc) in the current context.
+   This command is useful to remind the user of the names of library lemmas.
 
-   .. cmdv:: Search @string
+   .. todo above: "current context" meaning on the current goal or the first goal?
+      below: adapted the old wording, doesn't seem consistent
+      also will "invocation-of-tactics" cover the command case, or will there be a new section?
 
-      If :n:`@string` is a valid identifier, this command
-      displays the name and type of all objects (theorems, axioms, etc) of
-      the current context whose name contains string. If string is a
-      notation’s string denoting some reference :n:`@qualid` (referred to by its
-      main symbol as in `"+"` or by its notation’s string as in `"_ + _"` or
-      `"_ 'U' _"`, see Section :ref:`notations`), the command works like ``Search`` :n:`@qualid`.
+   * :n:`@string  {? % @ident }` - If :n:`@string` is a valid identifier,
+     search for objects whose name contains :n:`@string`. If :n:`@string` is a
+     notation's string for some reference :n:`@qualid` (referred to by its
+     main symbol as in `"+"` or by its notation’s string as in `"_ + _"` or
+     `"_ 'U' _"`, see Section :ref:`notations`) the command works like :cmd:`Search` :n:`@qualid`.
 
-   .. cmdv:: Search @string%@ident
+     .. todo: don't understand the 2nd sentence; notation's string -> notation string?
+        what if it's not a valid identifier?  error or does something different?
+        this is a substring match, e.g. "abc" matches "bc", right?
 
-      The string string must be a notation or the main
-      symbol of a notation which is then interpreted in the scope bound to
-      the delimiting key :token:`ident` (see Section :ref:`LocalInterpretationRulesForNotations`).
+     .. todo would be nice if :n: had syntax for embedding :cmd: et al
 
-   .. cmdv:: Search @term_pattern
+     Specifying :n:`@ident` limits this to interpretation in the scope bound to
+     the delimiting key :token:`ident` (see Section :ref:`LocalInterpretationRulesForNotations`).
 
-      This searches for all statements or types of
-      definition that contains a subterm that matches the pattern
-      :token:`term_pattern` (holes of the pattern are either denoted by `_` or by
-      :n:`?@ident` when non linear patterns are expected).
+     .. todo: does delimiting key mean anything special?
+
+   * :n:`@one_term` - Search for objects containing a subterm matching the pattern
+     :token:`one_term` in which holes of the pattern are indicated by `_` or,
+     when a non linear pattern is specified, :n:`?@ident`.
+
+      .. todo: "all statements or types of definition" is equivalent to "objects"?
+         Should non linear pattern be defined or reworded?  I had never run across it.
+         Not so keen on the term "holes", maybe reword to remove
```

Comment:
I don't understand the first part of the question. Non linear patterns mean some pattern in which the same variable appears several times, example:

```
Require Import Arith.
Search (?n + _ = _ + ?n).
(* Nat.add_comm: forall n m : nat, n + m = m + n *)
```

Reply:
Yes, I looked it up in Wikipedia.

2nd try: Should the terminology "non linear pattern" be defined here or elsewhere or replaced with the description such as "if the same variable occurs more than once in a pattern, all occurrences must match the same value".  I'm not sure using the term is helpful:
- I had not run across the term before, but the concept is pretty obvious.  Maybe it's a term specialists know and love.
- A pattern that doesn't have this property seems rather pitiful
- The term suggests 2 alternatives.  We don't use the term "linear pattern" AFAIK.
- "non linear" is not very suggestive of the property it describes

Your thoughts?

Reply:
I agree the term does not bring much and can be removed. All the user needs to know if that if they need a variable to be repeated in the pattern, they can use `?x` instead of `_`.